core/src/main/scala/org/apache/spark/internal/config/package.scala

```diff
@@ -2023,6 +2023,22 @@ package object config {
       .stringConf
       .createWithDefaultString("PWR")
 
+  private[spark] val EXECUTOR_DECOMMISSION_BATCH_INTERVAL =
+    ConfigBuilder("spark.executor.decommission.batchInterval")
```

Comment:
Is there any way to reasonably pick a good value for this configuration option?

The load issue is related to the amount of actual blocks being transferred and (indirectly) the number of executors currently being decommissioned. Interval based tuning can't produce stable throttling in this case, because it's not load-aware and would almost always decommission at a slower rate than the cluster could support, or at a higher rate than the cluster could handle.

It doesn't seem like this solves the problem described in practice.

What if, instead of batches, throttling was based on decommission concurrency? The number of active nodes participating in a decommission would be much closer proxy to the amount of load being produced, and would have the advantage of shutting down quicker when there was little work to do or slower if there is a lot of work to do (with the caveat that we'd be hurt by stalling or slow-boat decommission unlike in this approach)

Reply:
I like this idea, setting a maximum number of concurrent nodes for Spark commanded decommissioning . That being said, I think if the cluster manager / cloud provider triggers a SIGPWR or otherwise marks a node for decommissioning I think we shouldn't limit the number of decommissioning executors since the CM/cloud provider likely has their own time limits so we need to do what we can.

I know this probably complicates the code, if we want we could do this as a multi-part PR.